### PR TITLE
Query: Adds Index Metrics to Stream API

### LINKED ISSUE #2855: IndexMetrics on QueryResponse
Currently IndexMetrics is on QueryResponse<T> but not QueryResponse,

see: https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs

It would be great to have the new IndexMetrics feature on the QueryResponse object; is there a technical limitation why this cant be done? Currently I am using The 'Stream' response so I can handle deserialization without Newtonsoft Json, so QueryResponse<T> is not being used.

### PR DESCRIPTION
**Pull request by leminh98**
# Pull Request Template

## Description

This is a feature request from customers to make the Index Metrics field available to access from the stream API.

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #2855

### CODE REVIEW

**Review by sboshra — APPROVED** [review]
:shipit:

**Review by sboshra — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs [hunk 1]
```diff
@@ -57,6 +57,21 @@ internal QueryResponse()
             this.ResponseLengthBytes = responseLengthBytes;
             this.memoryStream = memoryStream;
             this.CosmosSerializationOptions = serializationOptions;
+
+            if (responseHeaders != null)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:61]: Should this logic be put on response message? If someone has a custom handler they would get they base64 string.

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs [hunk 2]
```diff
@@ -57,6 +57,21 @@ internal QueryResponse()
...
+            {
+                if (responseHeaders.IndexUtilizationText != null)
+                {
+                    this.IndexUtilizationText = new Lazy<string>(() =>
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:65]: Avoid duplicate code. Add a utility method.

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs [hunk 3]
```diff
@@ -182,14 +201,20 @@ class QueryResponse<T> : FeedResponse<T>
...
-            this.IndexUtilizationText = new Lazy<string>(() =>
+            if (responseMessageHeaders != null)
             {
-                IndexUtilizationInfo parsedIndexUtilizationInfo = IndexUtilizationInfo.CreateFromString(responseMessageHeaders.IndexUtilizationText);
-                StringBuilder stringBuilder = new StringBuilder();
-                IndexMetricWriter indexMetricWriter = new IndexMetricWriter(stringBuilder);
-                indexMetricWriter.WriteIndexMetrics(parsedIndexUtilizationInfo);
-                return stringBuilder.ToString();
-            });
+                if (responseMessageHeaders.IndexUtilizationText != null)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:206]: nit: 
```suggestion
                if (responseMessageHeaders?.IndexUtilizationText != null)
```

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs (hunk 1 above)
Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:61]: I think since this is not something that you want to turn on all the time, so I feel like CX shouldn't try and handle the base64 string on its own. Besides, the base 64 encoded json contains some extra info that might confuses cx even more if they try and understand it themselves

Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:61]: Oh, so you are saying to move this to the base class. Ah, that makes sense. I misunderstood you before. Yes, that makes sense I will look into moving it there

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs (hunk 2 above)
Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:65]: Not sure where the utility method would be since the two class are not really related in any shape or form.

Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:65]: Decided to go with internal static helper method on ResponseMessage

File: Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs [hunk 4]
```diff
@@ -81,6 +83,9 @@ public ResponseMessage()
             this.RequestMessage = requestMessage;
             this.CosmosException = cosmosException;
             this.Headers = headers ?? new Headers();
+
+            this.IndexUtilizationText = DecodeIndexMetrics(this.Headers);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs:87]: Prefix with class name

File: Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs [hunk 5]
```diff
@@ -237,5 +254,28 @@ private void CheckDisposed()
...
+                        IndexUtilizationInfo parsedIndexUtilizationInfo = IndexUtilizationInfo.CreateFromString(responseMessageHeaders.IndexUtilizationText);
+                        StringBuilder stringBuilder = new StringBuilder();
+                        IndexMetricWriter indexMetricWriter = new IndexMetricWriter(stringBuilder);
+                        indexMetricWriter.WriteIndexMetrics(parsedIndexUtilizationInfo);
+
+                        return stringBuilder.ToString();
+                    });
+            }
+
+            return new Lazy<string>();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs:278]: why not return null and avoid object creation?

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs [hunk 6]
```diff
@@ -182,14 +182,7 @@ class QueryResponse<T> : FeedResponse<T>
...
 
-            this.IndexUtilizationText = new Lazy<string>(() =>
-            {
-                IndexUtilizationInfo parsedIndexUtilizationInfo = IndexUtilizationInfo.CreateFromString(responseMessageHeaders.IndexUtilizationText);
-                StringBuilder stringBuilder = new StringBuilder();
-                IndexMetricWriter indexMetricWriter = new IndexMetricWriter(stringBuilder);
-                indexMetricWriter.WriteIndexMetrics(parsedIndexUtilizationInfo);
-                return stringBuilder.ToString();
-            });
+            this.IndexUtilizationText = ResponseMessage.DecodeIndexMetrics(responseMessageHeaders);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:185]: Why is this needed? Shouldn't the base class already handle this?

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs [hunk 7]
```diff
@@ -208,7 +201,7 @@ class QueryResponse<T> : FeedResponse<T>
 
         private Lazy<string> IndexUtilizationText { get; }
 
-        public override string IndexMetrics => this.IndexUtilizationText.Value;
+        public override string IndexMetrics => this.IndexUtilizationText?.Value;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:204]: No need to override it right?

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs (hunk 6 above)
Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:185]: this is the feed option one, not the response message

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs (hunk 7 above)
Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:204]: this is the one inheriting from FeedResponse, not the one inheriting from ResponseMessage. I already removed the one that inherited from ResponseMessage

File: Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs (hunk 5 above)
Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs:278]: good point. changed to null

File: Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs (hunk 4 above)
Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs:87]: do you mean to change it to:
`this.IndexUtilizationText = ResponseMessage.DecodeIndexMetrics(this.Headers);`

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs]: Yes

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs (hunk 6 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:185]: What if they access the headers directly? Should it return the formatted text or the byte array?

string t = responseMessage.IndexUtilization; //Formatted text
string x = responseMessage.Headers.IndexUtilization // Return the byte array

Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:185]: I think the best thing to do is to make the headers internal. People shouldn't access this header directly anyway, unless you see a scenario where that would be more preferable?

Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:185]: Actually I just checked. The IndexUtilizationText header is already internal, so people shouldn't be able to access the header directly.  They have to access it through responseMessage.IndexMetrics

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryResponse.cs:185]: What is the reasoning for exposing it on the ResponseMessage instead of the Headers?

**Review by sboshra — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs [hunk 8]
```diff
@@ -127,6 +132,18 @@ public virtual Stream Content
...
+
+        /// <summary>
+        /// Gets the Index Metrics in the current <see cref="ResponseMessage"/> to be used for debugging purposes. 
+        /// It's applicable to query response only. Other feed response will return null for this field.
+        /// This result is only available if QueryRequestOptions.PopulateIndexMetrics is set to true.
+        /// </summary>
+        /// <value>
+        /// The index utilization metrics.
+        /// </value>
+        public string IndexMetrics => this.IndexUtilizationText?.Value;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs:145]: With Track2 all stream API's will be replaced by the core SDK's Response Type. 
Extension methods as work-around impacted modern UT which is already reported extensively and not a recommended pattern. 

Also as a genral pattern not all of Typed contract is promoted at stream contract. 
Let's just start Typed API for now.

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs:145]: Update from offline sync-up with @j82w

Extension methods is not great but is a good alternative for such corner scenarios. So we can leverage this model in V4 if needed.

Signing it off now.

**Review by kirankumarkolli — APPROVED** [review]
